### PR TITLE
Async for `bittensor/extrinsics/network.py`

### PR DESCRIPTION
**Pull request by basfroman**
Part 2 of https://github.com/opentensor/bittensor/issues/1990

### CODE REVIEW

File: bittensor/extrinsics/network.py [hunk 1]
```diff
@@ -15,37 +15,44 @@
...
+
         success (bool):
-            Flag is ``true`` if extrinsic was finalized or included in the block.
-            If we did not wait for finalization / inclusion, the response is ``true``.
+            Flag is ``true`` if extrinsic was finalized or included in the block. If we did not wait for finalization / inclusion, the response is ``true``.
     """
-    your_balance = subtensor.get_balance(wallet.coldkeypub.ss58_address)
-    burn_cost = bittensor.utils.balance.Balance(subtensor.get_subnet_burn_cost())
+    your_balance = await subtensor.get_balance(wallet.coldkeypub.ss58_address)
+    burn_cost = balance.Balance(await subtensor.get_subnet_burn_cost())
```
Comment (thewhaleking) [inline bittensor/extrinsics/network.py]: This should be gathered.

File: bittensor/extrinsics/network.py [hunk 2]
```diff
@@ -62,48 +69,58 @@ def register_subnetwork_extrinsic(
...
+            wait_for_inclusion=wait_for_inclusion,
+            wait_for_finalization=wait_for_finalization,
+        )
+
+        # We only wait here if we expect finalization.
+        if not wait_for_finalization and not wait_for_inclusion:
+            return True
+
+        # process if registration successful
+        # TODO: ? passably have to be awaited after `submit_extrinsic` implemented in `async_substrate.py`
```
Comment (thewhaleking) [inline bittensor/extrinsics/network.py:96]: "Possibly". This is a good callout. Not yet finished with the extrinsics stuff in substrate.

File: bittensor/extrinsics/network.py [hunk 3]
```diff
@@ -167,46 +174,44 @@ def set_hyperparameter_extrinsic(
...
-            # Successful registration, final check for membership
-            else:
-                bittensor.__console__.print(
-                    f":white_heavy_check_mark: [green]Hyper parameter {parameter} changed to {value}[/green]"
+        # We only wait here if we expect finalization.
+        if not wait_for_finalization and not wait_for_inclusion:
+            return True
+
+        # process if registration successful
+        # TODO: ? passably have to be awaited after `submit_extrinsic` implemented in `async_substrate.py`
```
Comment (thewhaleking) [inline bittensor/extrinsics/network.py:206]: "Possibly"

File: bittensor/extrinsics/network.py [hunk 4]
```diff
@@ -167,46 +174,44 @@ def set_hyperparameter_extrinsic(
...
+        response.process_events()
+        if not response.is_success:
+            bittensor.__console__.print(
+                ":cross_mark: [red]Failed[/red]: error:{}".format(
+                    response.error_message
                 )
-                return True
+            )
+            await asyncio.sleep(0.5)
```
Comment (thewhaleking) [inline bittensor/extrinsics/network.py:214]: Why do we wait here? I know we do in the original also, but why?

File: bittensor/subtensor.py [hunk 5]
```diff
@@ -5106,7 +5106,7 @@ async def get_subnet_burn_cost(self, block: Optional[int] = None) -> Optional[st
         if lock_cost is None:
             return None
 
-        return lock_cost
+        return int(lock_cost)
```
Comment (thewhaleking) [inline bittensor/subtensor.py:5109]: The docs state this should be in Tao (which is usually float). Is it actually in Rao? Even if so, why are we changing the type here?

File: bittensor/utils/async_substrate.py [hunk 6]
```diff
@@ -344,7 +349,8 @@ async def get_block_number(self, block_hash: str) -> int:
                 return int(response["result"]["number"], 16)
 
     def close(self):
-        raise NotImplementedError()
+        self.substrate.close()
+        # TODO: Websocket close need to be implemented and added here.
```
Comment (thewhaleking) [inline bittensor/utils/async_substrate.py:353]: Is there a reason we want to manually allow closing the connection?

**Review by thewhaleking — CHANGES_REQUESTED** [review]

File: bittensor/extrinsics/network.py (hunk 1 above)
Comment (basfroman) [reply bittensor/extrinsics/network.py]: Good catch! Done

File: bittensor/extrinsics/network.py (hunk 4 above)
Comment (basfroman) [reply bittensor/extrinsics/network.py:214]: Honestly, I don't see any valuable reasons for this action. But I just convert this into async. Let's remove sleep at all?

File: bittensor/utils/async_substrate.py (hunk 6 above)
Comment (basfroman) [reply bittensor/utils/async_substrate.py:353]: For the same reasons as in original SubstrateInterface. Then it used in Subtensor.

**Review by thewhaleking — APPROVED** [review]

**Pull request merged**